src/cache.cc

```diff
@@ -74,162 +76,178 @@ void CACHE::handle_fill()
+		  }
+		else
+		  {
+		    if (MSHR.entry[mshr_index].instruction)
+		      upper_level_icache[fill_cpu]->return_data(&MSHR.entry[mshr_index]);
+		    if (MSHR.entry[mshr_index].is_data)
+		      upper_level_dcache[fill_cpu]->return_data(&MSHR.entry[mshr_index]);
+		  }
+	      }
 
-            // update processed packets
-            if (cache_type == IS_ITLB) { 
+	      // update processed packets
+	      if (cache_type == IS_ITLB) {
                 MSHR.entry[mshr_index].instruction_pa = block[set][way].data;
                 if (PROCESSED.occupancy < PROCESSED.SIZE)
-                    PROCESSED.add_queue(&MSHR.entry[mshr_index]);
-            }
-            else if (cache_type == IS_DTLB) {
+		  PROCESSED.add_queue(&MSHR.entry[mshr_index]);
+	      }
+	      else if (cache_type == IS_DTLB) {
                 MSHR.entry[mshr_index].data_pa = block[set][way].data;
                 if (PROCESSED.occupancy < PROCESSED.SIZE)
-                    PROCESSED.add_queue(&MSHR.entry[mshr_index]);
-            }
-            else if (cache_type == IS_L1I) {
+		  PROCESSED.add_queue(&MSHR.entry[mshr_index]);
+	      }
+	      else if (cache_type == IS_L1I) {
                 if (PROCESSED.occupancy < PROCESSED.SIZE)
-                    PROCESSED.add_queue(&MSHR.entry[mshr_index]);
-            }
-            //else if (cache_type == IS_L1D) {
-            else if ((cache_type == IS_L1D) && (MSHR.entry[mshr_index].type != PREFETCH)) {
+		  PROCESSED.add_queue(&MSHR.entry[mshr_index]);
+	      }
+	      //else if (cache_type == IS_L1D) {
+	      else if ((cache_type == IS_L1D) && (MSHR.entry[mshr_index].type != PREFETCH)) {
                 if (PROCESSED.occupancy < PROCESSED.SIZE)
-                    PROCESSED.add_queue(&MSHR.entry[mshr_index]);
-            }
+		  PROCESSED.add_queue(&MSHR.entry[mshr_index]);
+	      }
 
-	    if(warmup_complete[fill_cpu] && (MSHR.entry[mshr_index].cycle_enqueued != 0))
-	      {
-		uint64_t current_miss_latency = (current_core_cycle[fill_cpu] - MSHR.entry[mshr_index].cycle_enqueued);
-		/*
-		if(cache_type == IS_L1D)
-		  {
+	      if(warmup_complete[fill_cpu] && (MSHR.entry[mshr_index].cycle_enqueued != 0))
+		{
+		  uint64_t current_miss_latency = (current_core_cycle[fill_cpu] - MSHR.entry[mshr_index].cycle_enqueued);
+		  /*
+		    if(cache_type == IS_L1D)
+		    {
 		    cout << current_core_cycle[fill_cpu] << " - " << MSHR.entry[mshr_index].cycle_enqueued << " = " << current_miss_latency << " MSHR index: " << mshr_index << endl;
-		  }
-		*/
-		total_miss_latency += current_miss_latency;
-	      }
+		    }
+		  */
+		  total_miss_latency += current_miss_latency;
+		}
 	  
-            MSHR.remove_queue(&MSHR.entry[mshr_index]);
-            MSHR.num_returned--;
+	      MSHR.remove_queue(&MSHR.entry[mshr_index]);
+	      MSHR.num_returned--;
 
-            update_fill_cycle();
-        }
+	      writes_available_this_cycle--;
+
+	      update_fill_cycle();
+	    }
+	  }
     }
+    else
+      {
+	return;
```

Comment:
I don't like the `else-return` idiom here. I think it would be more clear to invert the condition on the `if` and have that return at the top of the function. A similar thing goes for the `if` statement that follows.